src/providers/ipa/ipa_access.c

```diff
@@ -680,11 +689,10 @@ static void ipa_pam_access_handler_done(struct tevent_req *subreq)
 
     ret = ipa_hbac_evaluate_rules(state->be_ctx,
                                   state->access_ctx->ipa_options, state->pd);
-    if (ret == EOK) {
-        state->pd->pam_status = PAM_SUCCESS;
-    } else if (ret == ERR_ACCESS_DENIED) {
+    /* Do not set pam_status in case of success. */
```

Comment:
Hi,

this might be a bit fragile since you cannot be sure that `pd->pam_status` is not changed in `ipa_hbac_evaluate_rules()` as well.

I would suggest to add a new member variable to `state` to hold some information about the previous SDAP check and use it if `ipa_hbac_evaluate_rules()` returns `EOK`. Since it currently only affects `ERR_PASSWORD_EXPIRED_RENEW`/`PAM_NEW_AUTHTOK_REQD` this member might be a boolean, e.g. password_change_required. As an alternative you can safe `pd->pam_status` in the new member, e.g. sdap_pam_status, before calling `ipa_fetch_hbac_send()` and re-set it here if the HBAC check was successful.

bye,
Sumit

Reply:
I implemented a simple version of your proposal.